core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -130,8 +134,19 @@
    * <p>If this {@code TransportSet} has been shut down, the returned future will have {@code null}
    * value.
    */
-  ListenableFuture<ClientTransport> obtainActiveTransport() {
-    return activeTransportFuture;
+  final ListenableFuture<ClientTransport> obtainActiveTransport() {
+    synchronized (lock) {
+      if (activeTransportFuture == null) {
+        createActiveTransportFuture();
+      }
+      return Preconditions.checkNotNull(activeTransportFuture, "activeTransportFuture");
+    }
+  }
+
+  // To be mocked in tests
+  @VisibleForTesting
+  long currentTimeMillis() {
+    return System.currentTimeMillis();
```

Comment:
Actually, this should really be `System.nanoTime()` since you want to calculate elapsed time, which would be `Stopwatch`.